### ISSUE TITLE
Off heap memory overhead when enabling profiling

### ISSUE THREAD

**Issue by andersenleo** [DataDog/dd-trace-java#5869]
Hi,
we've noticed a significant RSS memory usage for processes that runs with the APM profiling enabled.
For some services the overhead have been up to 60% (3Gb -> 5Gb). This extra memory consumption is only visible on the OS, not in the JVM metrics.

In all cases we've been able to get rid of the extra memory consumption by disabling profiling (`DD_PROFILING_ENABLED=false`).

We've reproduced the behaviour on dd-trace-java versions `1.14.0`, `1.19.2`, and on the latest version: `1.20.1~70cd67ce90`. 
We've used the following JDK version with the same result:

```
openjdk version "17.0.6-ea" 2023-01-17 LTS
OpenJDK Runtime Environment (Red_Hat-17.0.6.0.9-0.3.ea.el8) (build 17.0.6-ea+9-LTS)
OpenJDK 64-Bit Server VM (Red_Hat-17.0.6.0.9-0.3.ea.el8) (build 17.0.6-ea+9-LTS, mixed mode, sharing)
```

```
openjdk version "17.0.8.1" 2023-08-24
OpenJDK Runtime Environment Temurin-17.0.8.1+1 (build 17.0.8.1+1)
OpenJDK 64-Bit Server VM Temurin-17.0.8.1+1 (build 17.0.8.1+1, mixed mode, sharing)
```

This shows to runs of a service with profiling enabled, and then without profiling enabled:

![Screenshot 2023-09-12 at 09 19 11](https://github.com/DataDog/dd-trace-java/assets/21565302/079948d6-af4a-4c60-a579-eababfa6c0fb)

The config output from the agent at startup is:

(with profiling)

```json
{
  "version": "1.20.1~70cd67ce90",
  "os_name": "Linux",
  "os_version": "4.18.0-492.el8.x86_64",
  "architecture": "amd64",
  "lang": "jvm",
  "lang_version": "17.0.8.1",
  "jvm_vendor": "Eclipse Adoptium",
  "jvm_version": "17.0.8.1+1",
  "java_class_version": "61.0",
  "http_nonProxyHosts": "null",
  "http_proxyHost": "null",
  "enabled": true,
  "service": "bo",
  "agent_url": "http://localhost:8126",
  "agent_error": false,
  "debug": false,
  "trace_propagation_style_extract": [
    "datadog"
  ],
  "trace_propagation_style_inject": [
    "datadog"
  ],
  "analytics_enabled": false,
  "sampling_rules": [
    {},
    {}
  ],
  "priority_sampling_enabled": true,
  "logs_correlation_enabled": true,
  "profiling_enabled": true,
  "remote_config_enabled": true,
  "debugger_enabled": false,
  "appsec_enabled": "ENABLED_INACTIVE",
  "telemetry_enabled": true,
  "dd_version": "4.0.304",
  "health_checks_enabled": true,
  "configuration_file": "no config file present",
  "runtime_id": "4b9b1bb7-c327-4ac7-b40f-b6cbcd1645ea",
  "logging_settings": {
    "levelInBrackets": false,
    "dateTimeFormat": "'[dd.trace 'yyyy-MM-dd HH:mm:ss:SSS Z']'",
    "logFile": "System.err",
    "configurationFile": "simplelogger.properties",
    "showShortLogName": false,
    "showDateTime": true,
    "showLogName": true,
    "showThreadName": true,
    "defaultLogLevel": "INFO",
    "warnLevelString": "WARN",
    "embedException": false
  },
  "cws_enabled": false,
  "cws_tls_refresh": 5000,
  "datadog_profiler_enabled": true,
  "datadog_profiler_safe": true,
  "datadog_profiler_enabled_overridden": false

```

(without profiling):

```json
{
  "version": "1.20.1~70cd67ce90",
  "os_name": "Linux",
  "os_version": "4.18.0-492.el8.x86_64",
  "architecture": "amd64",
  "lang": "jvm",
  "lang_version": "17.0.8.1",
  "jvm_vendor": "Eclipse Adoptium",
  "jvm_version": "17.0.8.1+1",
  "java_class_version": "61.0",
  "http_nonProxyHosts": "null",
  "http_proxyHost": "null",
  "enabled": true,
  "service": "bo",
  "agent_url": "http://localhost:8126",
  "agent_error": false,
  "debug": false,
  "trace_propagation_style_extract": [
    "datadog"
  ],
  "trace_propagation_style_inject": [
    "datadog"
  ],
  "analytics_enabled": false,
  "sampling_rules": [
    {},
    {}
  ],
  "priority_sampling_enabled": true,
  "logs_correlation_enabled": true,
  "profiling_enabled": false,
  "remote_config_enabled": true,
  "debugger_enabled": false,
  "appsec_enabled": "ENABLED_INACTIVE",
  "telemetry_enabled": true,
  "dd_version": "4.0.304",
  "health_checks_enabled": true,
  "configuration_file": "no config file present",
  "runtime_id": "dea98d03-855b-43b0-bdc4-5258430966f5",
  "logging_settings": {
    "levelInBrackets": false,
    "dateTimeFormat": "'[dd.trace 'yyyy-MM-dd HH:mm:ss:SSS Z']'",
    "logFile": "System.err",
    "configurationFile": "simplelogger.properties",
    "showShortLogName": false,
    "showDateTime": true,
    "showLogName": true,
    "showThreadName": true,
    "defaultLogLevel": "INFO",
    "warnLevelString": "WARN",
    "embedException": false
  },
  "cws_enabled": false,
  "cws_tls_refresh": 5000,
  "datadog_profiler_enabled": true,
  "datadog_profiler_safe": true,
  "datadog_profiler_enabled_overridden": false
}
```

We've decided to disable profiling until we understand why this is happening.

**Comment by richardstartin** [comment]
Thanks for the report @andersenleo. If you could create a support ticket we can take a look at your profiles and figure out what the cause is. 

The most likely cause is the size of the metadata (class names, method names, method signatures, or potentially metadata from tracing used to annotate profile events) which is application-dependent, which is consistent with your observation that the overhead varies from service to service. Applications which use a lot of reflection, or otherwise generate large numbers of classes can have higher metadata overhead. 

In the next release we will record the sizes of various internal data structures so we can pinpoint issues like these, and we plan to save some space by normalising certain class name patterns before emitting a profile.

**Comment by andersenleo** [comment]
Hi, @richardstartin - I'll take a stab at a support ticket as well 👍.
The memory areas you mention - wouldn't they be visible on the off-heap are of the standard JVM Metrics? We didn't see any increase there at all, only when looking at the process at OS level.

**Comment by andersenleo** [comment]
Support ticket created: #1338215

**Comment by richardstartin** [comment]
> The memory areas you mention - wouldn't they be visible on the off-heap are of the standard JVM Metrics? We didn't see any increase there at all, only when looking at the process at OS level.

no - this only tells you about direct bytebuffers, whereas we allocate native memory in the profiler.

Thanks for creating the support ticket, I will follow up there, but leave this issue open in case others encounter similar behaviour.

**Comment by andersenleo** [comment]
Hi @richardstartin - our support ticket is closed now, but I would still like to test the dd-agent with the "fixes" once it becomes available.
Is there an ETA for this?

**Comment by richardstartin** [comment]
@andersenleo the potential fixes (normalisation of reflected frames at source) were released in 1.21.0 so you can try now.

I'll check what happened on the support ticket - you should be able to comment on it again to reopen it. I will be able to investigate further if the memory consumption is not reduced.

**Comment by andersenleo** [comment]
Thanks for the quick response :) 
Great - we'll plan a new run against the new version!

**Comment by andersenleo** [comment]
I had time do to a quick comparison of 1.19.2 and the latest 1.21.0. We started a stand-alone service instance with profiling enabled to see how RSS memory usage looked.

Kind of the opposite of what we expected ;-) 

<img width="538" alt="Screenshot 2023-10-12 at 15 50 28" src="https://github.com/DataDog/dd-trace-java/assets/21565302/90b2f92d-27fc-4369-bd4b-97dde47aa0a6">

The graph that sticks around 3Gb is the "normal" service instances (running without the profiler), and the two other graphs are the same service running with profiling (on 1.19.2 and 1.21.0, from left to right).

Before stopping each profiled process the RSS (and HWM) was:

**For 1.19.2**
```
[root@xxx]# cat /proc/4033541/status|grep RSS -B 1
VmHWM:	 3417936 kB
VmRSS:	 3417936 kB
```
**For 1.21.0**
```
[root@xxx]# cat /proc/4041245/status|grep RSS -B 1
VmHWM:	 4222812 kB
VmRSS:	 4222812 kB
```

The used RSS mem usage for the latest version was over the top.. 

We're starting the JVM (JDK 17) with:

```
export DD_ENV=xxx
export DD_PROFILING_ENABLED=true
export DD_SERVICE=xxx
export DD_SERVICE_MAPPING=kafka:xxx,mysql:xxx,rabbitmq:xxx
export DD_TAGS=owner:xxx
export DD_VERSION=<version>
export PORT_HTTP=40201

java-17 -XX:+DisableExplicitGC -Xlog:gc*:gc.log -XX:+UseShenandoahGC -javaagent:/opt/datadog-apm-agent/dd-java-agent-<version>.jar -XX:FlightRecorderOptions=stackdepth=256 -Xmx1536m -Xms1536m -XX:MaxMetaspaceSize=763m -Dserver.port=$PORT_HTTP -jar xxx.jar
```

Any thoughts on this? And what would you recommend that we try (wrt flags etc)?

**Comment by richardstartin** [comment]
hi @andersenleo this is obviously not a good outcome. It's clear that we don't understand the cause of the memory overhead you're seeing yet, so we will need to investigate further. However, we are about to release (early next week) changes in 1.22.0 which result in significant reduction in CPU overhead and peak RSS, which may help.

**Comment by richardstartin** [comment]
Could you try disabling a few flags to rule something out please?

1. `-Ddd.profiling.ddprof.enabled=false` - it looks like the reduction in metadata size was quite effective for the relevant service because the chunk recorded by this profiler is now ~10% of the size of the chunk recorded by JFR. If disabling this fixes the problem, it looks like we have a memory leak in native code and will need to pinpoint it with our native profiler which has memory leak detection.
2. `-Ddd.profiling.ultra.minimal=true` - this disables most usage of built in JFR features, and I can see that there are large numbers of various built in and Datadog custom events in your recordings. The test was also run with a early access build, so there may be JFR bugs at play.

**Comment by andersenleo** [comment]
Ran the service with the settings recommended. It looks much better (but I'm not sure we're seeing much profile data for the service).

![Screenshot 2023-10-13 at 09 09 50](https://github.com/DataDog/dd-trace-java/assets/21565302/02ff997d-9b06-4c97-b47d-43b9132a5107)

```
[root@xxxx]# cat /proc/75453/status|grep RSS -B 1
VmHWM:	 3057496 kB
VmRSS:	 3057496 kB
```

**Comment by richardstartin** [comment]
@andersenleo thanks for doing that. Sorry, I meant could you run with each flag separately, so we can blame one component or the other? If `-Ddd.profiling.ddprof.enabled=false` is the effective flag, the improvements I mentioned earlier will be available next week. If the improvements aren't effective, we can use our native profiler to debug. Otherwise it's likely a JFR problem and we'll need to follow up with OpenJDK.

**Comment by andersenleo** [comment]
Ah, got it (a bit busy here today) :) 
I'll try make another run later today!

**Comment by MariekeVandamme** [comment]
@andersenleo , @richardstartin , seems like this thread died out? Was there a final solution or other thread created for this? We are facing the same issue, and are very interested in the outcome. Because atm the only option we see is to disable APM for the services that are affected.
Thanks!

**Comment by richardstartin** [comment]
Hi @MariekeVandamme can you try setting some flags _individually_ so we can narrow down the cause, otherwise it's very difficult to debug outlier defect reports like this which aren't occurring in infrastructure we can directly observe and experiment with:
* `-Ddd.profiling.ddprof.enabled=false` - this disables one of the profilers where the memory leak may be occurring - if you still have a memory leak with this flag, it's in JFR or tracing.
*  `-Ddd.profiling.ultra.minimal=true` - switches off most built in JDK JFR events - if you still have a memory leak, it's in tracing or ddprof
* `-Ddd.profiling.enabled=true` - switches off all profiling - if you still have a memory leak, the problem is in tracing.

**Comment by richardstartin** [comment]
We released a fix for a very slow memory leak in [1.30.0](https://github.com/DataDog/dd-trace-java/releases) which should show up in the "Serviceability" section of NMT ([Native Memory Tracking](https://docs.oracle.com/javase/8/docs/technotes/guides/troubleshoot/tooldescr007.html)). This is more likely to be the cause of the leak if you have many or very long JVM arguments (the total number of bytes for JVM argument _keys_ is the explanatory factor). If you are running on Java 20 or later, we report [NMT JFR events](https://www.morling.dev/blog/tracking-java-native-memory-with-jdk-flight-recorder/) and are currently working on visualising these as a metric, which will make spotting native memory leaks (be it in our tooling, the JVM, your libraries, or your app) easier. 

I will close this issue as it is probably fixed, but please reopen it if the issue persists after 1.30.0.

**Comment by Sineaggi** [comment]
@richardstartin is there a ticket/milestone we can follow for the metric visualization of native memory?
